8302850: Consider implementing a C1 clone intrinsic that uses ArrayCopyNode for primitive arrays

galderz opened the pull request:
Adding C1 intrinsic for primitive array clone invocations for aarch64 and x86 architectures.

The intrinsic includes a change to avoid zeroing the newly allocated array because its contents are copied over within the same intrinsic with arraycopy. This means that the performance of primitive array clone exceeds that of primitive array copy. As an example, here are the microbenchmark results on darwin/aarch64:

```
$ make test TEST="micro:java.lang.ArrayClone" MICRO="JAVA_OPTIONS=-XX:TieredStopAtLevel=1"
Benchmark                 (size)  Mode  Cnt    Score    Error  Units
ArrayClone.byteArraycopy       0  avgt   15    3.476 ?  0.018  ns/op
ArrayClone.byteArraycopy      10  avgt   15    3.740 ?  0.017  ns/op
ArrayClone.byteArraycopy     100  avgt   15    7.124 ?  0.010  ns/op
ArrayClone.byteArraycopy    1000  avgt   15   39.301 ?  0.106  ns/op
ArrayClone.byteClone           0  avgt   15    3.478 ?  0.008  ns/op
ArrayClone.byteClone          10  avgt   15    3.562 ?  0.007  ns/op
ArrayClone.byteClone         100  avgt   15    5.888 ?  0.206  ns/op
ArrayClone.byteClone        1000  avgt   15   25.762 ?  0.203  ns/op
ArrayClone.intArraycopy        0  avgt   15    3.199 ?  0.016  ns/op
ArrayClone.intArraycopy       10  avgt   15    4.521 ?  0.008  ns/op
ArrayClone.intArraycopy      100  avgt   15   17.429 ?  0.039  ns/op
ArrayClone.intArraycopy     1000  avgt   15  178.432 ?  0.777  ns/op
ArrayClone.intClone            0  avgt   15    3.406 ?  0.016  ns/op
ArrayClone.intClone           10  avgt   15    4.272 ?  0.006  ns/op
ArrayClone.intClone          100  avgt   15   13.110 ?  0.122  ns/op
ArrayClone.intClone         1000  avgt   15  113.196 ? 13.400  ns/op
```

It also includes an optimization to avoid instantiating the array copy stub in scenarios like this.

I run hotspot compiler tests successfully limiting them to C1 compilation darwin/aarch64, linux/x86_64 and linux/686. E.g.

```
$ make test TEST="hotspot_compiler" JTREG="JAVA_OPTIONS=-XX:TieredStopAtLevel=1"
...
   TEST                                              TOTAL  PASS  FAIL ERROR
   jtreg:test/hotspot/jtreg:hotspot_compiler          1234  1234     0     0
```

One question I had is what to do about non-primitive object arrays, see my [question](https://bugs.openjdk.org/browse/JDK-8302850?focusedId=14634879&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-14634879) on the issue. @cl4es any thoughts?

Thanks @rwestrel for his help shaping this up :)

cl4es commented:
This looks very much like what I had in mind: implement a C1 clone intrinsic by reusing existing arraycopy code as much as possible while not going through too much trouble to do so. While typically the benefits will only be to startup/warmup, the OpenJDK itself and many others do run short running processes with `-XX:TieredStopAtLevel=1` where this might result in a significant overall speed-up. Nice work! 

Sorry for erroneously referencing C2's `ArrayCopyNode` in the RFE summary. Feel free to rephrase this as "Implement C1 clone intrinsic that reuses arraycopy code for primitive arrays"

Regarding your question on object arrays then I have no strong opinion except that if it's much more work then it's probably not worth it. It's less used than the primitive clones, e.g. `java.util.Array.copyOfRange(T[], int, int, Class<? extends T[]>)` does not use `clone` unlike its primitive array relatives. Recent changes to such APIs made use of primitive array clone more prominent, which in turn made the C1 cost difference between clone and arraycopy more apparent.

galderz commented:
Thanks @cl4es for the feedback and the additional information regarding primitive vs object array cloning. Happy to leave it at primitives. I've amended the issue description.

dean-long commented:
Rather than create a new Clone op that fuses NewTypeArray and ArrayCopy, but requires new platform-specific code, why can't we just build clone on top of the existing NewTypeArray and ArrayCopy?  We just need the new flag to avoid zeroing.

galderz commented:
@dean-long I hadn't considered that but I will definitely look into it. I see there's a c1 type for `NewTypeArray` but I don't see any for array copy. Do you have any pointers/examples on how to approach a solution like the one you suggest?

dean-long commented:
@galderz, let me see if I can get it to work.

dean-long commented:
I think the right solution would be to add a line in `GraphBuilder::build_graph_for_intrinsic` for _clone, to append the IR for NewTypeArray and ArrayCopy as if we parsed newarray and arraycopy() from bytecodes.  I'll see if I can get that working tomorrow.

galderz commented:
I've added a couple of commits to fine tune the current approach (I will still explore the other option Dean mentioned):

The first commit makes sure the clone intrinsic handles deoptimizations appropiately. `compiler/interpreter/Test6833129` was failing because of this.

I've also added a commit to add a null checks for the array being cloned. This also includes a jtreg test for the scenario in which the array was null after C1 compilation. This was resulting in a segv that was not reproduced by any other existing test.

galderz commented:
I've fixed all the tests with the solution with the current approach. I'm now looking into the alternative solution @dean-long proposed.

galderz commented:
@dean-long I think I've got something working but need to carry out tests on my side. It needs to be fine tuned but it does indeed look way simpler.

galderz commented:
I've just pushed a new solution implemented, as much as possible, at the graph level. I've reverted the previous solution to keep history linear.

I've run the hotspot compiler tests successfully on darwin/aarch64 and linux/x86:

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR
   jtreg:test/hotspot/jtreg:hotspot_compiler          1242  1242     0     0
==============================
TEST SUCCESS

Finished building target 'test' in configuration 'release-darwin-arm64'
```

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR
   jtreg:test/hotspot/jtreg:hotspot_compiler          1235  1235     0     0
==============================
TEST SUCCESS

Finished building target 'test' in configuration 'release-linux-x86_64'
```

The array clone microbenchmark performance numbers show the expected improvements in both setups:

```
Benchmark                 (size)  Mode  Cnt    Score   Error  Units
ArrayClone.byteArraycopy       0  avgt   15    3.467 ? 0.006  ns/op
ArrayClone.byteArraycopy      10  avgt   15    3.673 ? 0.047  ns/op
ArrayClone.byteArraycopy     100  avgt   15    7.255 ? 0.025  ns/op
ArrayClone.byteArraycopy    1000  avgt   15   39.119 ? 0.595  ns/op
ArrayClone.byteClone           0  avgt   15    2.996 ? 0.008  ns/op
ArrayClone.byteClone          10  avgt   15    3.117 ? 0.020  ns/op
ArrayClone.byteClone         100  avgt   15    5.487 ? 0.063  ns/op
ArrayClone.byteClone        1000  avgt   15   25.232 ? 0.445  ns/op
ArrayClone.intArraycopy        0  avgt   15    3.198 ? 0.029  ns/op
ArrayClone.intArraycopy       10  avgt   15    4.428 ? 0.065  ns/op
ArrayClone.intArraycopy      100  avgt   15   17.015 ? 0.150  ns/op
ArrayClone.intArraycopy     1000  avgt   15  176.464 ? 1.820  ns/op
ArrayClone.intClone            0  avgt   15    2.891 ? 0.019  ns/op
ArrayClone.intClone           10  avgt   15    3.794 ? 0.008  ns/op
ArrayClone.intClone          100  avgt   15   12.289 ? 0.084  ns/op
ArrayClone.intClone         1000  avgt   15  112.032 ? 3.787  ns/op
Finished running test 'micro:java.lang.ArrayClone'
Test report is stored in build/release-darwin-arm64/test-results/micro_java_lang_ArrayClone
```

```
Benchmark                 (size)  Mode  Cnt    Score    Error  Units
ArrayClone.byteArraycopy       0  avgt   15    7.209 ?  0.168  ns/op
ArrayClone.byteArraycopy      10  avgt   15    8.433 ?  0.154  ns/op
ArrayClone.byteArraycopy     100  avgt   15   13.175 ?  1.063  ns/op
ArrayClone.byteArraycopy    1000  avgt   15  143.002 ?  2.946  ns/op
ArrayClone.byteClone           0  avgt   15    6.909 ?  0.120  ns/op
ArrayClone.byteClone          10  avgt   15    8.129 ?  0.744  ns/op
ArrayClone.byteClone         100  avgt   15    9.196 ?  0.201  ns/op
ArrayClone.byteClone        1000  avgt   15   60.831 ?  1.147  ns/op
ArrayClone.intArraycopy        0  avgt   15    6.674 ?  0.137  ns/op
ArrayClone.intArraycopy       10  avgt   15    8.815 ?  0.176  ns/op
ArrayClone.intArraycopy      100  avgt   15   42.191 ?  1.376  ns/op
ArrayClone.intArraycopy     1000  avgt   15  553.157 ? 51.817  ns/op
ArrayClone.intClone            0  avgt   15    6.376 ?  0.130  ns/op
ArrayClone.intClone           10  avgt   15    7.690 ?  0.652  ns/op
ArrayClone.intClone          100  avgt   15   25.115 ?  0.483  ns/op
ArrayClone.intClone         1000  avgt   15  245.194 ? 17.418  ns/op
Finished running test 'micro:java.lang.ArrayClone'
Test report is stored in build/release-linux-x86_64/test-results/micro_java_lang_ArrayClone
```

galderz commented:
Some linux 686 tests failing due to frame map not being big enough (side effect of not instantiating `ArrayCopyStub` for clone like use cases). I've ported the fix from the old approach to the new making necessary adjustments. I ran hotspot compiler tests with only C1 successfully on linux/x86_64 and linux/686 with the fix. Let's see if CI agrees.

dean-long commented:
/reviewers 2

galderz commented:
@dean-long any chance you could have another look at this?

dean-long commented:
I'm looking at it again, and I'm trying to figure how we can minimize platform-specific changes.  I'm hoping we can move some of the set_force_reexecute boiler-plate code into shared code.  We probably don't need _force_reexecute in CodeEmitInfo anymore.

dean-long reviewed (changes requested):
Your front-end changes require back-end changes, which are only implemented for x86 and aarch64.  So you need a way to disable this for other platforms, or port the fix to all platforms.  Minimizing the amount of platform-specific code required would also help.

dean-long commented:
IR expansion in append_alloc_array_copy() looks unconditional.  What's going to happen on platforms with no back-end support?

galderz commented:
> IR expansion in append_alloc_array_copy() looks unconditional. What's going to happen on platforms with no back-end support?

I might be wrong but the way I understood the code, I think other platforms will have no issue with that with the way the current code works:

The check I added in `Compiler::is_intrinsic_supported()` means that for clone calls in other platforms it would return false. If that returns false, `AbstractCompiler::is_intrinsic_available` will return false. Then this means that in `GraphBuilder::try_inline_intrinsics` `is_available` would be false, in which case the method will always return false and `build_graph_for_intrinsic` will not be called. `GraphBuilder::append_alloc_array_copy` is called from `build_graph_for_intrinsic`, so I don't see a danger of that being called for non-supported platforms.

dean-long commented:
OK, I missed the is_intrinsic_supported change.  The platform-specific changes should probably have a comment saying they are for clone support.  Also, I was hoping there was a way to minimize platform-specific changes, maybe by handing the force_reexecute inheritance in state_for(), and putting the state in x->state() instead of x->state_before().

galderz commented:
> OK, I missed the is_intrinsic_supported change. The platform-specific changes should probably have a comment saying they are for clone support.

That's fine for me, but the platform-specific changes are a bit spread around. Anywhere in particular you'd want the comment(s) to go in? Or shall the comments be added in all platform specific changes?

> Also, I was hoping there was a way to minimize platform-specific changes, maybe by handing the force_reexecute inheritance in state_for()....

I'm not sure exactly what you mean, but let me try to explain the `x->state_before()` before bit. So I guess you're referring to this piece of logic:

```cpp
  CodeEmitInfo* info = nullptr;
  if (x->state_before() != nullptr && x->state_before()->force_reexecute()) {
    info = state_for(x, x->state_before());
    info->set_force_reexecute();
  } else {
    info = state_for(x, x->state());
  }
```

This code was added to deal with SEGV failures coming out of `compiler/interpreter/Test6833129.clone_and_verify`, which is configure to stress deoptimizations. In `append_alloc_array_copy`, I construct `NewTypeArray` and the array_copy `Intrinsic` both with `state_before` so that if any deoptimizations happen at either stage, re-execution happen from the point before clone was called.

I guess I can move the code up to `LIRGenerator::state_for` and if the intrinsic is either array copy or new type array, apply that logic, is that what you are after?

> ... putting the state in x->state() instead of x->state_before()

Hmmm, you mean this instead? 

```
  CodeEmitInfo* info = nullptr;
  if (x->state_before() != nullptr && x->state_before()->force_reexecute()) {
    info = state_for(x, x->state());
    info->set_force_reexecute();
  } else {
    info = state_for(x, x->state());
  }
```

Granted that it could be tidied but not sure I understand how state can work here. Before this change both new type array and array copy called `CodeEmitInfo* info = state_for(x, x->state());` which didn't seem to be enough to be able to go back to re-executing bytecode before clone was called. It seemed to leave things in a half-done state when deoptimizations happened while clone's intrinsic was doing either a new type array or array copy.

dean-long commented:
I only wanted the comments around the boilerplate force_reexecute() logic, but if you are happy with my idea to move that logic into LIRGenerator::state_for then the comment could go there.  If not, I may look at it in a follow-up RFE, because I would like to get rid of the force_reexecute() hack that I added and see if I can instead tie it to the use of state_before() or ValueStack::StateBefore.

dean-long commented on src/hotspot/share/c1/c1_GraphBuilder.cpp:
```diff
@@ -2133,11 +2133,23 @@ void GraphBuilder::invoke(Bytecodes::Code code) {
+      // or when the target of the clone is not a Phi node
+      ciType* receiver_type;
+      if (target->get_Method()->intrinsic_id() == vmIntrinsics::_clone &&
+          ((receiver_type = state()->stack_at(state()->stack_size() - inline_target->arg_size())->exact_type()) == nullptr || // clone target is phi
```
I don't think target-specific logic belongs here.  And I don't understand the point about Phi nodes.  Isn't the holder_known flag enough?  For primitive arrays, isn't it true that inline_target->get_Method()->intrinsic_id() == vmIntrinsics::_clone?

galderz replied:
> I don't think target-specific logic belongs here. And I don't understand the point about Phi nodes. Isn't the holder_known flag enough?

In my testing `holder_known` was not enough to detect objects that are not Phi. For example:

```
    static int[] test(int[] ints)
    {
        return ints.clone();
    }
```

`holder_known` is false when it tries to C1 compile `ints.clone()`, am I missing something here?

> For primitive arrays, isn't it true that inline_target->get_Method()->intrinsic_id() == vmIntrinsics::_clone?

Possibly, but in this part of the logic I'm trying to find situations in which I don't want to apply the `clone` intrinsic. And those situations are non-array objects, and for arrays, those whose elements are not primitives. I don't see how I can craft such a condition with only `inline_target->get_Method()->intrinsic_id() == vmIntrinsics::_clone`? IOW, that condition might be true for primitive arrays, but is it false for non-array objects and non-primitive arrays?

dean-long replied:
You're right about holder_known, but why do you need to check for _clone specifically at line 2137?  If there is logic missing that prevents an inlining attempt then I think it should be fixed first, rather than in a followup.

And I see that you need to do a receiver type check to allow only primitive arrays.  Can you do that in append_alloc_array_copy, and bailout if not successful?  The logic in build_graph_for_intrinsic would need to change slightly to support this.

dean-long replied:
I was able to remove the clone-specific logic in invoke() in two parts:

1. fix the type_is_exact logic to allow array receiver
2. move primitive array receiver check into append_alloc_array_copy

galderz replied:
> You're right about holder_known, but why do you need to check for _clone specifically at line 2137? If there is logic missing that prevents an inlining attempt then I think it should be fixed first, rather than in a followup.

I added that check because none of the conditions in that `if` statement satisfied the situations in which `clone` calls are optimized. For the example I gave above, `code == Bytecodes::_invokevirtual` is true and `target->is_final_method()` is false. So that's why I added `clone` specifically.

> I was able to remove the clone-specific logic in invoke() in two parts:
> 
> 1. fix the type_is_exact logic to allow array receiver
> 2. move primitive array receiver check into append_alloc_array_copy

Great! I assume you also solved the clone check in line 2137?

How do we add your work on top of mine? Do I cherry pick the commit(s) from a branch of yours? Or some other way?

galderz commented:
> I only wanted the comments around the boilerplate force_reexecute() logic, but if you are happy with my idea to move that logic into LIRGenerator::state_for then the comment could go there. If not, I may look at it in a follow-up RFE, because I would like to get rid of the force_reexecute() hack that I added and see if I can instead tie it to the use of state_before() or ValueStack::StateBefore.

It might be better handled as a follow-up. The reason for the boilerplate code to look the way it does is because the original code didn't set `set_force_reexecute`. So by shaping the code in that way, I was trying to limit the impact of my changes to the specific case of my clone intrinsic, where both state before and force reexecute are set for the new type array and array copy intrinsics.

dean-long commented:
My suggested cleanup is here:
https://github.com/dean-long/jdk/tree/pr/17667
Also, you'll need a 2nd review.

dean-long commented:
My patch still needs some work.

dean-long commented:
I think we could eventually relax the requirement that receiver_klass be loaded, at least for object arrays, but for simplicity my patch will follow the existing behavior.

galderz commented:
@dean-long I tried your patch and my test worked fine with it. I had one doubt about the following:

```
    Value recv = apop();
    apush(recv);
```

Wouldn't you prefer to just peek the top of the stack rather than pop it to push it again? I would expect the former to be cheaper to do?

> My patch still needs some work.

What else does it need?

dean-long commented:
> @dean-long I tried your patch and my test worked fine with it. I had one doubt about the following:
> 
> ```
>     Value recv = apop();
>     apush(recv);
> ```
> 
> Wouldn't you prefer to just peek the top of the stack rather than pop it to push it again? I would expect the former to be cheaper to do?

Sure, feel free to improve it.

> > My patch still needs some work.
> 
> What else does it need?

I fixed the issue I had, so it should be good to test out now.  Thanks.

galderz commented:
> Sure, feel free to improve it.

I've just commented in https://github.com/openjdk/jdk/pull/18642/files#r1560900520, what do you think?

> I fixed the issue I had, so it should be good to test out now. Thanks.

Ok, so how would we integrate your changes? Do I just merge the commits in your https://github.com/dean-long/jdk/tree/pr/17667 branch?

galderz commented:
@dean-long I've added your commits on top of the PR, and added an additional commit to include https://github.com/openjdk/jdk/pull/18642#discussion_r1560900520. I've run compiler tests limiting to C1 on x86_64 and things looked good. I didn't test other architectures because the additional changes were limited to the shared code.

lmesnik commented on test/hotspot/jtreg/compiler/c1/TestNullArrayClone.java:
```diff
@@ -0,0 +1,67 @@
+        try {
+            test(null);
+            System.out.println("Expected NullPointerException to be thrown");
+            System.exit(97);
```
Please throw RuntimeException instead of System.exit().

lmesnik commented on test/hotspot/jtreg/compiler/c1/TestNullArrayClone.java:
```diff
@@ -0,0 +1,67 @@
+
+public class TestNullArrayClone {
+    public static void main(String[] args)
+    {
```
please change style to use
public static void main(String[] args) { 
instead of
public static void main(String[] args)
{
in the test code.

galderz commented:
> Re: Also, I don't see GHA checks running. Are they enabled in your repo?

Sorry, I had disabled the workflows on my fork because I was close to running out of free resources last month. I didn't realise it would have an impact here. I've re-enabled them and they are running now. @lmesnik I've pushed a commit to address your comments.

galderz commented:
> It would be a good idea to ask arm/ppc/riscv/s390 port maintainers to test your changes.

Sounds good, I'll update here when I've done so.

galderz commented:
FAO @bulasevich @TheRealMDoerr @RealFYang @RealLucy I've created [JDK-8330472](https://bugs.openjdk.org/browse/JDK-8330472) to get the changes in this PR tested in arm/ppc/riscv/s390 architectures, to make sure no regressions are introduced.

TheRealMDoerr commented on src/hotspot/cpu/x86/c1_MacroAssembler_x86.cpp:
```diff
@@ -300,11 +300,13 @@ void C1_MacroAssembler::allocate_array(Register obj, Register len, Register t1,
-  int base_offset = align_up(base_offset_in_bytes, BytesPerWord);
-  initialize_body(obj, arr_size, base_offset, len_zero);
+  if (zero_array) {
+      const Register len_zero = len;
```
hotspot uses 2 spaces indentation.

TheRealMDoerr commented on src/hotspot/share/c1/c1_GraphBuilder.cpp:
```diff
@@ -3651,9 +3653,13 @@ void GraphBuilder::build_graph_for_intrinsic(ciMethod* callee, bool ignore_retur
   case vmIntrinsics::_getAndSetReference     : append_unsafe_get_and_set(callee, false); return;
   case vmIntrinsics::_getCharStringU         : append_char_access(callee, false); return;
   case vmIntrinsics::_putCharStringU         : append_char_access(callee, true); return;
+  case vmIntrinsicID::_clone                 : append_alloc_array_copy(callee); return;
```
Why do you use `vmIntrinsicID` while all others use `vmIntrinsics`?

TheRealMDoerr commented:
> FAO @bulasevich @TheRealMDoerr @RealFYang @RealLucy I've created [JDK-8330472](https://bugs.openjdk.org/browse/JDK-8330472) to port the changes here to arm/ppc/riscv/s390. Also, the changes in this PR have been in made in such way that they only affect architectures on which the intrinsic is implemented. Would you also be able to test the changes in this PR to make sure no regressions are introduced on these archs?

Thanks! I'll test it. We will need separate issues for these platforms because they are maintained by different people.

galderz commented:
@TheRealMDoerr I've pushed a couple of commits to address your comments.

TheRealMDoerr commented:
Thanks for cleaning this up! Tests have passed on PPC64.

RealFYang commented:
> FAO @bulasevich @TheRealMDoerr @RealFYang @RealLucy I've created [JDK-8330472](https://bugs.openjdk.org/browse/JDK-8330472) to port the changes here to arm/ppc/riscv/s390. Also, the changes in this PR have been in made in such way that they only affect architectures on which the intrinsic is implemented. Would you also be able to test the changes in this PR to make sure no regressions are introduced on these archs?

FYI: This also test good on linux-riscv64 platform. Thanks for the ping!

bulasevich commented on src/hotspot/share/c1/c1_GraphBuilder.cpp:
```diff
@@ -4428,6 +4437,45 @@ void GraphBuilder::append_char_access(ciMethod* callee, bool is_store) {
 }
 
+void GraphBuilder::append_alloc_array_copy(ciMethod* callee) {
+  {
```
There are stylistic questions to this block
- why separate block?
- not necessary comment "Peek at receiver"
- misprint in "not primtive array" comment
- extra line for {
- excessive comment "// not primitive array"
- bailout message do not mention phi
- please use INLINE_BAILOUT macro

Comment about phi does not make things clear: not evident why receiver type can be nullptr and why it is phi. After all, why can't we check src->exact_type()->as_array_klass()->element_type() here?

galderz replied:
> There are stylistic questions to this block
...
> Comment about phi does not make things clear: not evident why receiver type can be nullptr and why it is phi. After all, why can't we check src->exact_type()->as_array_klass()->element_type() here?

This block comes from @dean-long, I'll let him comment.

dean-long replied:
> why separate block?

It's left over from when I had it wrapped with `if (UseNewCode)` as I was testing it.  It can be removed.

> not necessary comment "Peek at receiver"

OK.

> misprint in "not primtive array" comment

Good catch.  You can include the suggested fix using the github UI and get added as a contributor if you like.

> extra line for {

To me it makes it more readable, but if it goes against the style guide, feel free to change it.

> excessive comment "// not primitive array"

OK.

> bailout message do not mention phi

Yes, if we bailout due to `receiver_type == nullptr` then we should have a better bailout message.  But you may be correct that it is impossible.  This code was copied from an earlier version where it did seem possible.  If we change it we'll need a new round of testing.

> please use INLINE_BAILOUT macro

We would need a version of INLINE_BAILOUT that doesn't return false.

bulasevich commented on test/hotspot/jtreg/compiler/c1/TestNullArrayClone.java:
```diff
@@ -0,0 +1,62 @@
+
+/*
+ * @test
+ * @bug 8302850
```
please add @summary
is the purpose of the test to check that array clone throws NPE for null input and does not throw otherwise?
Don't we want to check the contents of the copied data?
Don't we want to check different sizes and array types?
Is 1K iterations enough to compile the method?

galderz replied:
> please add @summary

+1

> is the purpose of the test to check that array clone throws NPE for null input and does not throw otherwise?

I added this test to verify that null values for the array where handled correctly when the array clone call had been C1 compiled. I added this test at the time because I discovered a bug in the implementation and I had not seen any existing tests fail.

> Don't we want to check the contents of the copied data?
> Don't we want to check different sizes and array types?

Verifying the contents sounds good.

Different sizes and types sounds good as well, but what sizes would you choose? For the types, I would limit it to primitive types since the intrinsic is only implemented for those.

> Is 1K iterations enough to compile the method?

That seemed to be enough in my case to trigger the issue, what number should I use instead? I found `Tier3CompileThreshold` that is 2000, so maybe increase it to that just in case? Is there a way to verify from the test that the method has been C1 compiled before going ahead and invoking the method with null?

galderz replied:
Added brief test summary.

bulasevich replied:
Thank you!

> what sizes would you choose? For the types, I would limit it to primitive types

Yes, checking the primitive types is fine. Let it be int[], long[], and byte[]. For size I would pick an odd.

```
static final int ITER = 2000; // ~ Tier3CompileThreshold
static final int ARRAY_SIZE = 999;
```

bulasevich replied:
If you like, you can inspect the output of the -XX:+PrintLIR option to see if C1 applies arraycopy as expected.

galderz replied:
Expanded the tests in this class. I verified with `PrintLIR` that the methods are being C1 compiled.

bulasevich commented:
```
$ make test TEST="hotspot_compiler" JTREG="JAVA_OPTIONS=-XX:TieredStopAtLevel=1"
...
   TEST                                              TOTAL  PASS  FAIL ERROR
   jtreg:test/hotspot/jtreg:hotspot_compiler          1234  1234     0     0
```

1234 tests does not seem to be enough for a low-level feature. I would also check :hotspot_gc :hotspot_serviceability :hotspot_runtime and jdk tier1-3 targets.

offamitkumar commented:
Tests result is clean on s390x as well. 

```
command: make test TEST="hotspot_compiler" JTREG="JAVA_OPTIONS=-XX:TieredStopAtLevel=1"


==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/hotspot/jtreg:hotspot_compiler          1167  1167     0     0   
==============================
TEST SUCCESS
```

```
command: make run-test-tier1

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
>> jtreg:test/hotspot/jtreg:tier1                     2153  2151     2     0 <<
>> jtreg:test/jdk:tier1                               2359  2357     2     0 <<
   jtreg:test/langtools:tier1                         4477  4477     0     0   
   jtreg:test/jaxp:tier1                                 0     0     0     0   
   jtreg:test/lib-test:tier1                            33    33     0     0   
==============================
TEST FAILURE
```

The failures in `tier1` were not related to this PR. Let me know if more testing is required. 
CC:@reallucy

galderz commented:
Thanks for testing @offamitkumar @RealFYang 😀

galderz commented:
^ Thanks @TheRealMDoerr too :)

bulasevich commented on src/hotspot/share/c1/c1_GraphBuilder.cpp:
```diff
@@ -4428,6 +4437,45 @@ void GraphBuilder::append_char_access(ciMethod* callee, bool is_store) {
 }
 
+void GraphBuilder::append_alloc_array_copy(ciMethod* callee) {
+  {
+    // Peek at receiver
+    Value recv = state()->stack_at(state()->stack_size() - callee->arg_size());
+    ciType* receiver_type = recv->exact_type();
+    if (receiver_type == nullptr || // clone target is phi
+        !receiver_type->is_type_array_klass()) // not primtive array
+    {
+      // not primitive array
+      inline_bailout("clone array not primitive");
+      return;
+    }
+  }
```
```suggestion
  const int args_base = state()->stack_size() - callee->arg_size();
  ciType* receiver_type = state()->stack_at(args_base)->exact_type();
  if (receiver_type == nullptr) {
    inline_bailout("must have a receiver");
    return;    
  }
  if (!receiver_type->is_type_array_klass()) {
    inline_bailout("clone array not primitive");
    return;
  }
```

galderz commented:
> ```
> $ make test TEST="hotspot_compiler" JTREG="JAVA_OPTIONS=-XX:TieredStopAtLevel=1"
> ...
>    TEST                                              TOTAL  PASS  FAIL ERROR
>    jtreg:test/hotspot/jtreg:hotspot_compiler          1234  1234     0     0
> ```
> 
> 1234 tests does not seem to be enough for a low-level feature. I would also check :hotspot_gc :hotspot_serviceability :hotspot_runtime and jdk tier1-3 targets.

Fair point. I went ahead and run the following tests on linux/x64:

```
$ make test TEST="hotspot_compiler hotspot_gc hotspot_serviceability hotspot_runtime tier1 tier2 tier3" JTREG="JAVA_OPTIONS=-XX:TieredStopAtLevel=1 -Xcomp"
```

And got some failures:

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR
   jtreg:test/hotspot/jtreg:hotspot_compiler          1247  1247     0     0
   jtreg:test/hotspot/jtreg:hotspot_gc                 512   512     0     0
   jtreg:test/hotspot/jtreg:hotspot_serviceability     340   340     0     0
   jtreg:test/hotspot/jtreg:hotspot_runtime            793   793     0     0
   jtreg:test/hotspot/jtreg:tier1                     2050  2050     0     0
>> jtreg:test/jdk:tier1                               2369  2368     1     0 <<
   jtreg:test/langtools:tier1                         4477  4477     0     0
   jtreg:test/jaxp:tier1                                 0     0     0     0
   jtreg:test/lib-test:tier1                            33    33     0     0
   jtreg:test/hotspot/jtreg:tier2                      649   649     0     0
>> jtreg:test/jdk:tier2                               4100  4098     1     1 <<
>> jtreg:test/langtools:tier2                           11    10     1     0 <<
   jtreg:test/jaxp:tier2                               515   515     0     0
   jtreg:test/hotspot/jtreg:tier3                      245   245     0     0
>> jtreg:test/jdk:tier3                               1413  1406     4     3 <<
   jtreg:test/langtools:tier3                            0     0     0     0
   jtreg:test/jaxp:tier3                                 0     0     0     0
==============================
TEST FAILURE
```

On closer inspection, the tests that failed are:

```
jtreg:test/jdk/jdk/jfr/event/compiler/TestCompilerCompile.java
jtreg:test/jdk/java/security/AccessController/DoPrivAccompliceTest.java
jtreg:test/langtools/jdk/jshell/FailOverDirectExecutionControlTest.java
jtreg:test/jdk/java/rmi/server/RemoteServer/AddrInUse.java
jtreg:test/jdk/java/rmi/server/RMISocketFactory/useSocketFactory/unicast/TCPEndpointReadBug.java
jtreg:test/jdk/jdk/jfr/event/gc/stacktrace/TestMetaspaceParallelGCAllocationPendingStackTrace.java
jtreg:test/jdk/jdk/jfr/event/gc/stacktrace/TestMetaspaceSerialGCAllocationPendingStackTrace.java
```

I tried with the commit before my changes (https://github.com/openjdk/jdk/commit/eebcc218) and the tests also fail there when run with `JTREG="JAVA_OPTIONS=-XX:TieredStopAtLevel=1 -Xcomp"`

galderz commented:
/label rfr

rwestrel commented on src/hotspot/share/c1/c1_GraphBuilder.cpp:
```diff
@@ -2027,8 +2027,10 @@ void GraphBuilder::invoke(Bytecodes::Code code) {
       ciType* type = receiver->exact_type();
-      if (type != nullptr && type->is_loaded() &&
-          type->is_instance_klass() && !type->as_instance_klass()->is_interface()) {
+      if (type != nullptr && type->is_loaded()) {
```
Is it the case that we can't see an interface here? Or that we think it's ok if we see an interface here?

dean-long replied:
We can't see an interface here because it will get rejected by `ciInstanceKlass::exact_klass`, so we could even assert for that here if we wanted.

rwestrel replied:
Then, I think we should add an assert that `!type->as_instance_klass()->is_interface()` and also that it's not and array of interfaces (using `base_element_klass()`)

dean-long replied:
An array of interfaces can be exact:
```
    new Interface[20].getClasss();
```
and it seems like it would be safe to allow this, so I think we only need one assert for `!type->as_instance_klass()->is_interface()` if we don't trust the result of exact_type().

rwestrel replied:
Right. Then I think it would be safer to add an assert for `!type->as_instance_klass()->is_interface()`.

galderz replied:
@dean-long @rwestrel I've added the assert.

galderz replied:
The assert doesn't hold, e.g.

```
=== Output from failing command(s) repeated here ===
* For target buildtools_create_symbols_javac__the.COMPILE_CREATE_SYMBOLS_batch:
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/runner/work/jdk/jdk/src/hotspot/share/c1/c1_GraphBuilder.cpp:2031), pid=75212, tid=75244
#  Error: assert(!!type->as_instance_klass()->is_interface()) failed
#
# JRE version: OpenJDK Runtime Environment (23.0) (fastdebug build 23-internal-galderz-306db7459b1316251e36d0eccc3035d11db44889)
# Java VM: OpenJDK 64-Bit Server VM (fastdebug 23-internal-galderz-306db7459b1316251e36d0eccc3035d11db44889, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x777cd0]  GraphBuilder::invoke(Bytecodes::Code)+0x1200
```

Thoughts @rwestrel @dean-long?

galderz replied:
Hmmm, the double `!!`... let me fix that and see.

galderz replied:
Hmmm, something else is failing now. That's odd, maybe master has updated and is causing this PR to fail now?

```
#  Internal Error (/Users/runner/work/jdk/jdk/src/hotspot/share/ci/ciMetadata.hpp:88), pid=79328, tid=27395
#  assert(is_instance_klass()) failed: bad cast
```

I will look into it.

galderz replied:
Ah no, that assert comes from `type->as_instance_klass()` call:

```
  ciInstanceKlass*         as_instance_klass() {
    assert(is_instance_klass(), "bad cast");
    return (ciInstanceKlass*)this;
  }
```

@rwestrel @dean-long what shall we do here? Do we remove the assert altogether? Does the code need to change for the assert to pass? Any other ideas?

rwestrel commented on src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:
```diff
@@ -2504,7 +2505,9 @@ void LIR_Assembler::emit_arraycopy(LIR_OpArrayCopy* op) {
  }
 
-  __ bind(*stub->continuation());
+  if (stub != nullptr) {
```
Isn't that left over code from your previous implementation? Can `stub` be null?

rwestrel reviewed (commented):
Looks good overall otherwise.

galderz commented on src/hotspot/share/c1/c1_GraphBuilder.cpp:
```diff
@@ -2027,8 +2027,11 @@ void GraphBuilder::invoke(Bytecodes::Code code) {
-      if (type != nullptr && type->is_loaded() &&
-          type->is_instance_klass() && !type->as_instance_klass()->is_interface()) {
+      if (type != nullptr && type->is_loaded()) {
+        assert(!type->as_instance_klass()->is_interface(), "");
```
Thanks @dean-long for the suggested fix. CI looks good now.

rwestrel commented on src/hotspot/share/c1/c1_GraphBuilder.cpp:
```diff
@@ -2027,8 +2027,11 @@ void GraphBuilder::invoke(Bytecodes::Code code) {
-      if (type != nullptr && type->is_loaded() &&
-          type->is_instance_klass() && !type->as_instance_klass()->is_interface()) {
+      if (type != nullptr && type->is_loaded()) {
+        assert(!type->is_instance_klass() || !type->as_instance_klass()->is_interface(), "");
```
Please add a message to the assert.

galderz replied:
Added, is that ok?

rwestrel reviewed (approved):
Otherwise, looks good to me.

galderz commented:
/integrate

rwestrel commented:
/sponsor

galderz commented:
Thanks all for the reviews!
